bpo-40334: PEP 617: New PEG parser for CPython

pablogsal opened the pull request:
The SC has approved  PEP 617 so we'd like to land before alpha 6 goes out. We need reviews NOW.

<!-- issue-number: [bpo-40334](https://bugs.python.org/issue40334) -->
https://bugs.python.org/issue40334
<!-- /issue-number -->

pablogsal committed:
Run CI in pegen branch as well (#3)

gvanrossum committed:
Merge steps 1-3 and 6 into 'pegen' branch (#8)

- The pegen parser generator and its tests are in Tools/peg_generator. This is mostly copied from the https://github.com/gvanrossum/pegen repo, with some adaptations. This is still targeting 3.8.
- The new grammar (adapted to Python 3.9) is in Grammar/python.gram.
- There's a new builtin module peg_parser, configured in Setup.
- There's a new Make target regen-pegen which runs the parser generator with output directed to Modules/peg_parser/parse.c.

There's some code duplication between Tools/peg_generator and Modules/peg_parser, due to the 3.8/3.9 split and some tweaks needed to link with the core. Eventually this probably should go away.

lysnikolaou committed:
Fall back to a set for done in collect_todo (#9)

lysnikolaou committed:
Add Github Actions for pegen tests in Tools/peg_generator (#12)

lysnikolaou committed:
Allow expressions in yield_expression rule and rename it to annotated_rhs (#13)

gvanrossum committed:
Copy .clang-format from pegen project, to guide 'make format-c'

lysnikolaou committed:
Expose a C API and a Python-level module for pegen (#11)

* Use trailer meta for a function that calls the start rule
* Expose four C API functions that return either a mod_ty or a PyCodeObject *
* Expose two Python-level functions that return an AST object
* Extension module and C API does not get generated, but is hand-written
* Rewrite of run_parser* functions

Co-authored-by: Pablo Galindo <pablogsal@gmail.com>

gvanrossum committed:
Run clang-format over peg_parser/pegen.c

gvanrossum committed:
Run clang-format over Parser/pegen/pegen.c

gvanrossum committed:
Run clang-format over Modules/peg_parser.c

gvanrossum committed:
Ifdef-out token_name() to silence compiler output noise warning

gvanrossum committed:
Merge branch 'master' into pegen

lysnikolaou committed:
Make Ubuntu tests pass; disable Win and Mac tests (#19)

lysnikolaou committed:
Rewrite run_parser* functions to always return a mod_ty (#17)

* The caller is responsible for transforming the mod_ty a code or an AST object
* The C API functions don't expect the filename and the arena (where possible) objects as an
   argument, but construct them themselves

lysnikolaou committed:
Rewrite exported function (#21)

* Rewrite exported function:
    * It now initializes the keywords field so that the keywords variable can be static
    * It also accepts an argument, in order to be able to run the parser with different start rules
* Use the newly created function in fstring_compile_expr

lysnikolaou committed:
Merge remote-tracking branch 'upstream/master' into pegen

lysnikolaou committed:
Remove all unnecessary code from Tools/peg_generator; only keep the generator itself (#23)

* Remove C code from Tools/peg_generator/peg_parser and rely on Parser/pegen
* Move all tests to Lib/test/test_peg_generator/*
* Disable pegen tests on Github Actions

pablogsal committed:
Support for PEP 614 (#24)

lysnikolaou committed:
Re-enable MacOS test on Github Actions (#26)

gvanrossum committed:
Add memoization control (#28)

This is a straightforward port of https://github.com/gvanrossum/pegen/pull/236 .

In addition, slightly touched up .gitignore, Makefile (black), mypy.ini.

gvanrossum committed:
Fold long lines in metagrammar

lysnikolaou committed:
Update Parser/pegen/parse.c after addition of memoization flag (#28)

lysnikolaou committed:
Add a command-line flag and an environment variable to enable pegen (#29)

lysnikolaou committed:
Add missing possible start rules in python.gram (#32)

* Remove start rule and add file, interactive, eval and fstring rules
* Accept a mode parameter in C API functions

pablogsal committed:
Merge remote-tracking branch 'upstream/master' into pegen

pablogsal committed:
Update grammar to account for _Py_keyword changes (#40)

pablogsal committed:
Merge remote-tracking branch 'upstream/master' into pegen

lysnikolaou committed:
Add new C API functions PyPegen_*FromFileObject and use them in interactive mode (#38)

* Refactor raise_syntax_error to correctly handle interactive mode
* Correctly handle Ctrl-D in interactive mode

pablogsal committed:
Memoize some rules to avoid exponential time in deep nesting (#39)

Additionally, fix the script to calculate the nesting level using the generated extension.

lysnikolaou committed:
Run tests with pegen enabled on CI (#41)

* Fix audit tests
* Try to fallback to get_error_line if we cannot get the error line using PyErr_ProgramTextObject
* Add -E flag to run_tests.py only if PYTHONPARSER is not set

Co-authored-by: Pablo Galindo <pablogsal@gmail.com>

lysnikolaou committed:
Improve SyntaxError message when an ERRORTOKEN is returned by the tokenizer (#42)

Co-authored-by: Pablo Galindo <Pablogsal@gmail.com>

lysnikolaou committed:
Use FILE pointer instead of filename string to pass file to parser (#52)

When ./python is invoked with a file, which also encompasses
the case `echo "pass" | ./python", and the new parser is enabled,
pass the FILE * to the file to pegen, instead of the filename string
so that the above case is handled corretly.

lysnikolaou committed:
Add start_mark variable in loop rules (#51)

This solves a bug due to which a SEGFAULT occurs on parsing `a = 'sd`.
Note that the string is not terminated.

Code within the RHS of a loop needs to update the local mark variable,
which is then passed to the `insert_token` function, while its value
has been changed. But `insert_token` expects the mark of the starting
token and could under some circumstances SEGFAULT, when it gets
an invalidated `mark` parameter.

gvanrossum committed:
Simplify grammar rules around ASYNC (#53)

lysnikolaou committed:
Throw SyntaxError when parsing a STRING throws an exception (#54)

When parsing a STRING literal, UnicodeError's and ValueError's should
be caught and transformed into SyntaxError's.

lysnikolaou committed:
Add use_peg field to sys.flags (#55)

pablogsal committed:
Improve error reporting and disallow f(**kwargs, *args) (#46)

pablogsal committed:
Propagate errors from the tokenizer (#57)

lysnikolaou committed:
Output more descriptive error messages for named expressions (#56)

lysnikolaou committed:
Output more descriptive error messages for invalid assignments (#59)

lysnikolaou committed:
Avoid memory leak in Parser_New (#62)

lysnikolaou committed:
Fix SEGFAULT in test_fstring (#61)

In `fstring_compile_expr` the parser was being invoked by calling
`parse` directly, instead of `run_parser`. Thus, the jmp_buf wasn't
being initialized, leading to an invalid jump, upon calling `longjmp`.

lysnikolaou committed:
Provide location info on SyntaxError due to EOLS or EOFS errors (#60)

The current parser outputs a position pointer, when an EOLS or
EOFS error occurs, while parsing. With this, we imitate this
behavior:
>>> a = 'ex
  File "<stdin>", line 1
    a = 'ex
             ^
SyntaxError: EOL while scanning string literal

lysnikolaou committed:
Merge remote-tracking branch 'upstream/master' into pegen

lysnikolaou committed:
Ignore encoding cookie when that's necessary (#68)

lysnikolaou committed:
Silence compiler warnings about unused variables on Linux (#70)

lysnikolaou committed:
Merge remote-tracking branch 'upstream/master' into pegen

lysnikolaou committed:
Refactor illegal syntax rules to avoid invalid error messages (#72)

lysnikolaou committed:
Create a new rule for STRING+ and memoize it (#67)

Fixes #64.

lysnikolaou committed:
Fix type of strings rule (#73)

lysnikolaou committed:
Skip one test_traceback test, where pegen is better (#74)

lysnikolaou committed:
Skip non-critical test_string_literals test that fails (#75)

lysnikolaou committed:
Add script that reproduces all the PEP benchmarks (#71)

lysnikolaou committed:
Merge remote-tracking branch 'upstream/master' into pegen

lysnikolaou committed:
Fix calls to _PyInterpreterState_GET after merge

lysnikolaou committed:
Generate specialized error message for invalid string prefixes (#77)

Closes #66.

pablogsal committed:
Compress data/xxl.py and decompress on demand (#79)

pablogsal committed:
Use the new parser by default (#81)

lysnikolaou committed:
Rename pegen functions to start with the _PyPegen_ prefix (#80)

pablogsal committed:
Compile pegen in windows (#82)

* Compile pegen in windows

* fixup! Compile pegen in windows

* Use Py_ssize_t instead of int

* Try to fix permission errors

* Correct pegen_interface location

* fixup! Correct pegen_interface location

* Supress permission error exceptions

* fixup! fixup! fixup! Use the new parser by default

* fixup! fixup! fixup! fixup! Use the new parser by default

Co-authored-by: Lysandros Nikolaou <lisandrosnik@gmail.com>

pablogsal committed:
Remove whitespace to make patchcheck happy (#87)

pablogsal committed:
Rename the peg_parser module to _peg_parser (#86)

pablogsal committed:
Fix travis errors (#88)

tirkarthi commented on Doc/using/cmdline.rst:
```diff
@@ -252,6 +252,12 @@ Miscellaneous options
    options).  See also :envvar:`PYTHONDEBUG`.
 
 
+.. cmdoption:: -p old|new
```
versionadded tag can be added for new option.

tirkarthi commented on Doc/using/cmdline.rst:
```diff
@@ -574,6 +580,13 @@ conflict.
    :option:`-d` multiple times.
 
 
+.. envvar:: PYTHONPARSER
```
versionadded tag can be added for new environment variable.

lysnikolaou committed:
Remove the -p command line option and add a new -X oldparser option instead (#92)

lysnikolaou committed:
Disallow trailing comma in from-imports (#97)

gvanrossum committed:
Add blurb

pablogsal committed:
Use pegen in the symtable and fix symtable test errors (#102)

gvanrossum committed:
Merge branch 'master' into pegen

gvanrossum committed:
Only optimize __debug__ when ctx==Load (#106)

This avoids a segfault if an assignment to __debug__ makes it into the AST.

(However, it does not enough for #98 to close it.)

lysnikolaou committed:
Reproduce current parser's error messages for indentation errors (#107)

pablogsal committed:
Add more tokenize errors (#105)

pablogsal committed:
Normalize unicode identifiers (#108)

lysnikolaou committed:
Raise IndentationError on indentation-related errors (#111)

gvanrossum committed:
Copy most fixes from #96 except for the part that uses pegen for compile()

gvanrossum committed:
Temporarily suppress two failures in test_eof.py (one even with old parser!)

lysnikolaou committed:
Use col_offset when raising an E_DEDENT error (#116)

vstinner commented on .github/workflows/build.yml:
```diff
@@ -13,6 +13,7 @@ on:
     - '**/*.rst'
   pull_request:
     branches:
+    - pegen
```
This change should be reverted, no?

vstinner commented on .github/workflows/build.yml:
```diff
@@ -50,6 +51,22 @@ jobs:
+    - name: Display build info
+      run: make pythoninfo
+    - name: Tests
+      run: make buildbottest TESTOPTS="-j4 -uall,-cpu"
```
Is it really worth it to slow down the pre-commit to test the old parser on different platforms? A single pre-commit job would be enough, no? (Pick your favorite platform.) Or maybe only post-commit buildbots would be enough.

vstinner commented on .travis.yml:
```diff
@@ -1,5 +1,5 @@
 language: c
-dist: xenial
+dist: bionic
```
I would prefer to change the Travis CI base image in a separated PR.

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -426,6 +426,8 @@ Miscellaneous options
 
    * ``-X faulthandler`` to enable :mod:`faulthandler`;
+   * ``-X oldparser``: enable the traditional LL(1) parser.  See also
+     :envvar:`PYTHONOLDPARSER`.
```
Maybe the PEP 617 should be mentioned here. I suggest to directly deprecate this option and announces its removal in 3.10 using a `.. deprecated` markup.

Moreover, the new option should be documented in the ".. versionchanged:: 3.9" below.

Same remark for PYTHONOLDPARSER.

vstinner commented on Include/cpython/initconfig.h:
```diff
@@ -147,6 +147,10 @@ typedef struct {
 
+    /* Enable PEG parser?
+       1 by default, set to 0 by -X oldparser and PYTHONOLDPARSER */
+    int use_peg;
```
I suggest to rename the option to "_use_peg_parser". If you are not used to parsers, it's hard to guess "peg" meaning. I prefer to add "_" prefix to make it private since it's a temporary option for Python 3.9, will be removed from 3.10.

The new member should be documented at https://docs.python.org/dev/c-api/init_config.html#pyconfig with a ".. versionadded:: 3.9" markup. Again, I suggest to directly deprecate it and announce its removal in 3.10 using ".. deprecated-removed" markup.

vstinner commented on Include/pegen_interface.h:
```diff
@@ -0,0 +1,32 @@
+#ifndef Py_LIMITED_API
```
If the whole content is excluded from the limited C API, you can move the file to Include/cpython/ which means that you have to use `#include "cpython/pegen_interface.h"`.

lysnikolaou replied:
Do we need to have the `#ifndef Py_LIMITED_API` in that case?

vstinner commented on Python/sysmodule.c:
```diff
@@ -2427,6 +2427,7 @@ static PyStructSequence_Field flags_fields[] = {
     {"inspect",                 "-i"},
     {"interactive",             "-i"},
     {"optimize",                "-O or -OO"},
+    {"use_peg",                 "-p old or -p new"},
```
I would prefer to rename this field `_use_peg_parser`: repeat parser to make it easier to understand and make it private since it's going to be removed in Python 3.10.

`-p old or -p new` is outdated.

I would prefer to add the new field a the end.

lysnikolaou replied:
Done.

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+    if (!PyUnicode_IS_ASCII(id))
+    {
+        PyObject *id2;
+        if (!p->normalize && !init_normalization(p))
```
You might move `!p->normalize` test inside init_normalization(): do nothing (return 1) if p->normalize is not NULL.

lysnikolaou replied:
Done (Just commenting, in order to mark this as resolved)

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+                         "%.200s",
+                         _PyType_Name(Py_TYPE(id2)));
+            Py_DECREF(id2);
+            longjmp(p->error_env, 1);
```
Please add a comment to justify the usage of setjmp/longjmp. Maybe repeat the comment where setjmp() is called.

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+static inline Py_ssize_t
+byte_offset_to_character_offset(PyObject *line, int col_offset)
+{
+    const char *str = PyUnicode_AsUTF8(line);
```
I suggest to check str:

```
if (!str) {
    return 0;
}
```

lysnikolaou replied:
Done

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+#include <Python.h>
```
Would you mind to add `Parser/pegen` subdirectory to `SRCDIRS` in `configure.ac`? So `make tags` will also parse `Parser/pegen/*.c` and `Parser/pegen/*.h`.

lysnikolaou replied:
Done.

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+
+static void
+raise_decode_error(Parser *p)
+{
```
It's unclear to me if this function must be called with an exception set or not. It seems like an exception must be set. But then why tokenizer_error() exits with -1 if an exception is set, but call raise_decode_errorraise_decode_error() when no exception is set?

I'm concerned by PyObject_Str(value) below: can value be NULL?

Maybe add `assert(PyErr_Occurred());`.

lysnikolaou replied:
Done.

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+    PyErr_Fetch(&type, &value, &tback);
+    errstr = PyObject_Str(value);
+
+    Py_INCREF(Py_None);
+    PyObject *tmp = Py_BuildValue("(OiiN)", filename, 0, -1, Py_None);
```
```suggestion
    PyObject *tmp = Py_BuildValue("(OiiO)", filename, 0, -1, Py_None);
```

lysnikolaou replied:
Done.

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+
+    if (p->fill == p->size) {
+        int newsize = p->size * 2;
+        p->tokens = PyMem_Realloc(p->tokens, newsize * sizeof(Token *));
```
If PyMem_Realloc() fails, p->tokens is set to NULL and you leak memory.

lysnikolaou replied:
Done.

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+            return -1;
+        }
+        for (int i = p->size; i < newsize; i++) {
+            p->tokens[i] = PyMem_Malloc(sizeof(Token));
```
You should check that the memory allocation doesn't fail.

lysnikolaou replied:
Done.

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+    t->end_col_offset = end_col_offset;
+
+    // if (p->fill % 100 == 0) fprintf(stderr, "Filled at %d: %s \"%s\"\n", p->fill,
+    // token_name(type), PyBytes_AsString(t->bytes));
```
Maybe remove this debug code, or add an #ifdef? Same remark for commented fprintf below.

PyConfig has a parser_debug member. I don't know who use it :-)

lysnikolaou replied:
Just removed them.

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+                memo_statistics[type] += count;
+            }
+            p->mark = m->mark;
+            *(void **)(pres) = m->node;
```
pres is declared as `void*` but used as `void**`. That's surprising. Should pres be declared as `void**` instead?

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+{
+    Parser *p = PyMem_Malloc(sizeof(Parser));
+    if (p == NULL) {
+        PyErr_Format(PyExc_MemoryError, "Out of memory for Parser");
```
I suggest to use PyErr_NoMemory() instead which works even if there is no available memory, whereas PyErr_Format() allocates memory and so can fail as well.

lysnikolaou replied:
Done.

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+        PyErr_Format(PyExc_MemoryError, "Out of memory for tokens");
+        return NULL;
+    }
+    p->tokens[0] = PyMem_Malloc(sizeof(Token));
```
PyMem_Malloc() can fail.

lysnikolaou replied:
Done.

vstinner commented on Parser/pegen/pegen.c:
```diff
@@ -0,0 +1,1834 @@
+                number_of_dots += 1;
+                break;
+            default:
+                assert(current_expr->type == ELLIPSIS || current_expr->type == DOT);
```
Hum, maybe use Py_UNREACHABLE()?

lysnikolaou replied:
Done.

lysnikolaou committed:
Use pegen in compile and other functions in pythonrun.c (#96)

This revealed lots of other issues that we had to fix.

gvanrossum reviewed (commented):
@vstinner, which of your suggestions MUST be merged before we land this? We can iterate after alpha6, but I don't want to hold up @ambv and I do want to get this into alpha6, which is scheduled for today. (@ambv has promised to wait, but I don't want to try his patience too much.)

My proposal is to do everything post alpha6 except for the CI config changes. Right now we're scrambling on a fix for the memory leak caused by setjmp (https://github.com/we-like-parsers/cpython/pull/119).

pablogsal committed:
Don't use longjmp and check for the error indicator on every rule (#119)

* Don't use longjmp and check for the error indicator on every rule

* Address feedback

* Add missing semicolon

* PEP 7 fix

* Add missing goto and re-run black
